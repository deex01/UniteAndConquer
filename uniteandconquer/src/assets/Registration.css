.registration-page {
  margin: auto;
  height: 100%;
}
 
.submit {
  border: none;
  background: #FFCB7D;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  border-radius: 5px;
  padding: 15px;
  width: 60%;
 
  font-style: italic;
  font-weight: bold;
  font-size: 15px;
}
 
.registration-tags {
  background-color: #FFCB7D;
  height: 100%;
  max-width: 400px;
}
 
.tags {
  display: grid;
  grid-template-columns: 0.5fr 0.5fr;
}
 
.tag {
  background: #FFAB2D;
  border: 5px solid #FFFFFF;
  box-sizing: border-box;
  border-radius: 40px;
  text-align: center;
  width: 75px;
  height: 35px;
  margin: auto;
  margin-bottom: 20px;
}
 
.headers {
  color: white;
  padding: 6px;
  font-style: italic;
  text-align: center;
}
 
.label {
  text-align: left;
}
 
h3 {
  font-size: 30px;
}
 
.registration {
  height: 100%;
  display: grid;
  grid-template-columns: 0.3fr 0.7fr;
}
 
.registration-input {
  margin: auto;
  font-size: 20px;
  font-style: italic;
  text-align: center;
}
 
input {
  border: none;
  background-color: #FFD9A0;
  margin: auto;
  width: 90%;
  height: 30px;
  padding: 10px;
  font-size: 20px;
}
 
.registration-field {
  margin: 10px;
}
 
.submit:hover {
  box-shadow: 5px 5px 5px rgba(0, 0, 0, 0.25);
  background-color: grey;
  transition-duration: 0.4s;
}
 
.registration-text-fields {
  display: grid;
  grid-template-columns: 300px 300px;
}
 
.reg-bottom {
  text-align: center;
}

.input-reqs {
  background: rgba(172, 170, 170, 0.49);
  border-radius: 10px;
  width: 60%;
  padding-top: 1px;
  padding-left: 10px;
  padding-right: 10px;
  padding-bottom: 10px;
  text-align: left;
  margin: auto;
}

.rule {
  margin-left: 25px;
  font-size: 15px;
}

::placeholder {
  font-size: 5px;
}