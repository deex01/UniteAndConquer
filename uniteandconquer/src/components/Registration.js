/* eslint-disable no-unused-vars */
import React, { useState, useEffect } from 'react';
import { Link, useNavigate } from 'react-router-dom';

import '../assets/App.css';
import '../assets/Registration.css';

const UserDB = require('../modules/UserDB');

function Registration() {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [phone, setPhone] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [validFirstName, setValidFirstName] = useState(false);
  const [validLastName, setValidLastName] = useState(false);
  const [validPassword, setValidPassword] = useState(false);
  const [validPhone, setValidPhone] = useState(false);
  const [validEmail, setValidEmail] = useState(false);
  const [validReqs, setValidReqs] = useState(false);
  const [validConfirmPassword, setValidConfirmPassword] = useState(false);
  const [tags, setTags] = useState([]);
  const navigate = useNavigate();

  // register the user given the information provided
  // if the password and confimPassword are not correct, then throw an exception
  const registerUser = () => {
    if (password !== confirmPassword) {
      throw new Error('password and confirmPassword need to be the same');
    }
    UserDB.createUser(
      phone,
      email,
      password,
      firstName,
      lastName,
      tags,
      (success, id, err) => {
        if (success) {
          navigate('/login');
        } else {
          console.log(err);
        }
      },
    );
  };

  const firstNameValidation = () => {
    const alpha = /^([a-zA-Z]){2,15}$/;
    if (firstName) {
      if (firstName.match(alpha)) {
        setValidFirstName(true);
      } else {
        setValidFirstName(false);
      }
    }
  };

  const LastNameValidation = () => {
    const alpha = /^([a-zA-Z]){2,15}$/;

    if (lastName) {
      if (lastName.match(alpha)) {
        setValidLastName(true);
      } else {
        setValidLastName(false);
      }
    }
  };

  const phoneValidation = () => {
    const phoneRegex = /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/im;

    if (phone) {
      if (phone.match(phoneRegex)) {
        setValidPhone(true);
      } else {
        setValidPhone(false);
      }
    }
  };

  const emailValidation = () => {
    const emailRegex = /\S+@\S+\.\S+/;

    if (email) {
      if (email.match(emailRegex)) {
        setValidEmail(true);
      } else {
        setValidEmail(false);
      }
    }
  };

  const passwordValidation = () => {
    const pwRule = /^(?=.*?[A-Z])(?=(.*[a-z]){1,})(?=(.*[\d]){1,})(?=(.*[\W]){1,})(?!.*\s).{8,}$/;

    if (password) {
      if (password.match(pwRule)) {
        setValidPassword(true);
      } else {
        setValidPassword(false);
      }
    }
  };

  const confirmPasswordValidation = () => {
    if (password && password === confirmPassword) {
      setValidConfirmPassword(true);
    } else {
      setValidConfirmPassword(false);
    }
  };

  const allReqs = () => {
    if (validFirstName && validLastName && validConfirmPassword && validEmail && validPhone) {
      setValidReqs(true);
    } else {
      setValidReqs(false);
    }
  };

  useEffect(() => {
    firstNameValidation();
    LastNameValidation();
    passwordValidation();
    confirmPasswordValidation();
    phoneValidation();
    emailValidation();
    allReqs();
  });

  return (
    <div className="registration-page">
      <div className="registration">
        <div className="registration-tags">
          <div className="headers">
            <h1>Unite and Conquer</h1>
            <h3>Choose your interests</h3>
          </div>
          <div className="tags">
            <div className="tag">Tag1</div>
            <div className="tag">Tag2</div>
            <div className="tag">Tag3</div>
            <div className="tag">Tag4</div>
            <div className="tag">Tag5</div>
            <div className="tag">Tag6</div>
          </div>
        </div>
        <div className="registration-input">
          <h3>Get Started</h3>
          <div className="input-reqs">
            <div>
              <h2>Password Requires</h2>
              <div className="rule">1. at least 8 characters</div>
              <div className="rule">2. at least one special character (e.g. ! @ # ? ])</div>
              <div className="rule">3. at least one uppercase and one lowercase letter</div>
              <div className="rule">4. at least one number</div>
            </div>
            <div>
              <h2>First/Last Name Requires</h2>
              <div className="rule">1. 2-15 characters</div>
              <div className="rule">2. Alphabetic characters</div>
            </div>
          </div>
          <div className="registration-text-fields">
            <div className="registration-field">
              <div className="label">first name</div>
              <input data-testid="first-name-input" className="first-name-input" onChange={(e) => setFirstName(e.target.value)} />
              {validFirstName ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
            <div className="registration-field">
              <div className="label">last name</div>
              <input data-testid="last-name-input" className="last-name-input" onChange={(e) => setLastName(e.target.value)} />
              {validLastName ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
            <div className="registration-field">
              <div className="label">phone</div>
              <input data-testid="phone-input" className="phone-input" placeholder="### - ### - ####" onChange={(e) => setPhone(e.target.value)} />
              {validPhone ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
            <div className="registration-field">
              <div className="label">email</div>
              <input data-testid="email-input" className="email-input" placeholder="username@domain.com" onChange={(e) => setEmail(e.target.value)} />
              {validEmail ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
            <div className="registration-field">
              <div className="label">password</div>
              <input data-testid="password-input" className="password-input" onChange={(e) => setPassword(e.target.value)} />
              {validPassword ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
            <div className="registration-field">
              <div className="label">confirm password</div>
              <input data-testid="confirm-password-input" className="confirm-password-input" onChange={(e) => setConfirmPassword(e.target.value)} />
              {validConfirmPassword ? <i className="fas fa-check" /> : <i className="fas fa-times" />}
            </div>
          </div>
          <br />
          <div className="reg-bottom">
            <button disabled={!validReqs} className="submit" type="button" onClick={registerUser}>
              Register
            </button>
            <p>
              Already have an account?
              {' '}
              <Link to="/login">Login</Link>
              {' '}
            </p>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Registration;
