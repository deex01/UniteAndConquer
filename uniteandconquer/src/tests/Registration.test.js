/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import Registration from '../components/Registration';

// https://stackoverflow.com/questions/66284286/react-jest-mock-usenavigate

const mockedUsedLink = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: () => mockedUsedLink,
  useNavigate: () => mockedUsedNavigate,
}));

test('registration header', () => {
  render(<Registration />);
  const headerElement = screen.getByText('Get Started');
  expect(headerElement).toBeInTheDocument();
});

test('registration requirements', () => {
  render(<Registration />);
  const pwElement = screen.getByText('Password Requires');
  expect(pwElement).toBeInTheDocument();
  const nameElement = screen.getByText('First/Last Name Requires');
  expect(nameElement).toBeInTheDocument();
});

// first name

test('input valid first name', () => {
  render(<Registration />);
  const input = screen.getByTestId('first-name-input');
  fireEvent.change(input, { target: { value: 'John' } });
  expect(input.value).toBe('John');
});

test('input invalid first name', () => {
  render(<Registration />);
  const input = screen.getByTestId('first-name-input');
  fireEvent.change(input, { target: { value: '#@' } });
  expect(input.value).toBe('#@');
});

// last name

test('input valid last name', () => {
  render(<Registration />);
  const input = screen.getByTestId('last-name-input');
  fireEvent.change(input, { target: { value: 'Doe' } });
  expect(input.value).toBe('Doe');
});

test('input invalid last name', () => {
  render(<Registration />);
  const input = screen.getByTestId('last-name-input');
  fireEvent.change(input, { target: { value: '#@' } });
  expect(input.value).toBe('#@');
});

// email

test('input valid email', () => {
  render(<Registration />);
  const input = screen.getByTestId('email-input');
  fireEvent.change(input, { target: { value: 'john@gmail.com' } });
  expect(input.value).toBe('john@gmail.com');
});

test('input invalid email', () => {
  render(<Registration />);
  const input = screen.getByTestId('email-input');
  fireEvent.change(input, { target: { value: 'johndoe' } });
  expect(input.value).toBe('johndoe');
});

// phone

test('input valid phone', () => {
  render(<Registration />);
  const input = screen.getByTestId('phone-input');
  fireEvent.change(input, { target: { value: '234-234-2345' } });
  expect(input.value).toBe('234-234-2345');
});

test('input invalid phone', () => {
  render(<Registration />);
  const input = screen.getByTestId('phone-input');
  fireEvent.change(input, { target: { value: '23432' } });
  expect(input.value).toBe('23432');
});

// password

test('input valid password', () => {
  render(<Registration />);
  const input = screen.getByTestId('password-input');
  fireEvent.change(input, { target: { value: 'aAb123#abc' } });
  expect(input.value).toBe('aAb123#abc');
});

test('input invalid password', () => {
  render(<Registration />);
  const input = screen.getByTestId('password-input');
  fireEvent.change(input, { target: { value: '234' } });
  expect(input.value).toBe('234');
});

// confirm password

test('input valid confirm password', () => {
  render(<Registration />);
  const password = screen.getByTestId('password-input');
  fireEvent.change(password, { target: { value: 'aAb123#abc' } });
  expect(password.value).toBe('aAb123#abc');
  const confirm = screen.getByTestId('confirm-password-input');
  fireEvent.change(confirm, { target: { value: 'aAb123#abc' } });
  expect(confirm.value).toBe('aAb123#abc');
});

// finish registration (complete)

test('complete registration', () => {
  render(<Registration />);
  const firstName = screen.getByTestId('first-name-input');
  fireEvent.change(firstName, { target: { value: 'John' } });
  expect(firstName.value).toBe('John');
  const lastName = screen.getByTestId('last-name-input');
  fireEvent.change(lastName, { target: { value: 'Doe' } });
  expect(lastName.value).toBe('Doe');
  const email = screen.getByTestId('email-input');
  fireEvent.change(email, { target: { value: 'john@gmail.com' } });
  expect(email.value).toBe('john@gmail.com');
  const phone = screen.getByTestId('phone-input');
  fireEvent.change(phone, { target: { value: '234-234-2345' } });
  expect(phone.value).toBe('234-234-2345');
  const password = screen.getByTestId('password-input');
  fireEvent.change(password, { target: { value: 'aAb123#abc' } });
  expect(password.value).toBe('aAb123#abc');
  const confirm = screen.getByTestId('confirm-password-input');
  fireEvent.change(confirm, { target: { value: 'aAb123#abc' } });
  expect(confirm.value).toBe('aAb123#abc');
  fireEvent.click(screen.getByText('Register'));
  // const linkElement = screen.getByText('Notifications');
  // expect(linkElement).toBeInTheDocument();
});

test('snapshot registration', () => {
  const component = renderer.create(<Registration />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
