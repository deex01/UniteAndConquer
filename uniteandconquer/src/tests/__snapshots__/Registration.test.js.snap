// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot registration 1`] = `
<div
  className="registration-page"
>
  <div
    className="registration"
  >
    <div
      className="registration-tags"
    >
      <div
        className="headers"
      >
        <h1>
          Unite and Conquer
        </h1>
        <h3>
          Choose your interests
        </h3>
      </div>
      <div
        className="tags"
      >
        <div
          className="tag"
        >
          Tag1
        </div>
        <div
          className="tag"
        >
          Tag2
        </div>
        <div
          className="tag"
        >
          Tag3
        </div>
        <div
          className="tag"
        >
          Tag4
        </div>
        <div
          className="tag"
        >
          Tag5
        </div>
        <div
          className="tag"
        >
          Tag6
        </div>
      </div>
    </div>
    <div
      className="registration-input"
    >
      <h3>
        Get Started
      </h3>
      <div
        className="input-reqs"
      >
        <div>
          <h2>
            Password Requires
          </h2>
          <div
            className="rule"
          >
            1. at least 8 characters
          </div>
          <div
            className="rule"
          >
            2. at least one special character (e.g. ! @ # ? ])
          </div>
          <div
            className="rule"
          >
            3. at least one uppercase and one lowercase letter
          </div>
          <div
            className="rule"
          >
            4. at least one number
          </div>
        </div>
        <div>
          <h2>
            First/Last Name Requires
          </h2>
          <div
            className="rule"
          >
            1. 2-15 characters
          </div>
          <div
            className="rule"
          >
            2. Alphabetic characters
          </div>
        </div>
      </div>
      <div
        className="registration-text-fields"
      >
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            first name
          </div>
          <input
            className="first-name-input"
            data-testid="first-name-input"
            onChange={[Function]}
          />
          <i
            className="fas fa-times"
          />
        </div>
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            last name
          </div>
          <input
            className="last-name-input"
            data-testid="last-name-input"
            onChange={[Function]}
          />
          <i
            className="fas fa-times"
          />
        </div>
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            phone
          </div>
          <input
            className="phone-input"
            data-testid="phone-input"
            onChange={[Function]}
            placeholder="### - ### - ####"
          />
          <i
            className="fas fa-times"
          />
        </div>
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            email
          </div>
          <input
            className="email-input"
            data-testid="email-input"
            onChange={[Function]}
            placeholder="username@domain.com"
          />
          <i
            className="fas fa-times"
          />
        </div>
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            password
          </div>
          <input
            className="password-input"
            data-testid="password-input"
            onChange={[Function]}
          />
          <i
            className="fas fa-times"
          />
        </div>
        <div
          className="registration-field"
        >
          <div
            className="label"
          >
            confirm password
          </div>
          <input
            className="confirm-password-input"
            data-testid="confirm-password-input"
            onChange={[Function]}
          />
          <i
            className="fas fa-times"
          />
        </div>
      </div>
      <br />
      <div
        className="reg-bottom"
      >
        <button
          className="submit"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Register
        </button>
        <p>
          Already have an account?
           
           
        </p>
      </div>
    </div>
  </div>
</div>
`;
